src/ray/raylet/scheduling/policy/scheduling_policy.h

```diff
@@ -17,24 +17,86 @@
 #include <vector>
 
 #include "ray/raylet/scheduling/cluster_resource_data.h"
+#include "ray/raylet/scheduling/policy/scheduling_context.h"
 #include "ray/raylet/scheduling/policy/scheduling_options.h"
 #include "ray/raylet/scheduling/scheduling_ids.h"
 
 namespace ray {
 namespace raylet_scheduling_policy {
 
+// Status of resource scheduling result.
+struct SchedulingResultStatus {
+  bool IsFailed() const { return code == SchedulingResultStatusCode::FAILED; }
+  bool IsInfeasible() const { return code == SchedulingResultStatusCode::INFEASIBLE; }
+  bool IsSuccess() const { return code == SchedulingResultStatusCode::SUCCESS; }
+
+  enum class SchedulingResultStatusCode {
+    // Scheduling failed but retryable.
+    FAILED = 0,
+    // Scheduling failed and non-retryable.
+    INFEASIBLE = 1,
+    // Scheduling successful.
+    SUCCESS = 2,
+  };
+  SchedulingResultStatusCode code = SchedulingResultStatusCode::SUCCESS;
+};
+
+struct SchedulingResult {
+  static SchedulingResult Infeasible() {
+    SchedulingResult result;
+    result.status.code = SchedulingResultStatus::SchedulingResultStatusCode::INFEASIBLE;
+    return result;
+  }
+
+  static SchedulingResult Failed() {
+    SchedulingResult result;
+    result.status.code = SchedulingResultStatus::SchedulingResultStatusCode::FAILED;
+    return result;
+  }
+
+  static SchedulingResult Success(std::vector<scheduling::NodeID> &&nodes) {
+    SchedulingResult result;
+    result.status.code = SchedulingResultStatus::SchedulingResultStatusCode::SUCCESS;
+    result.selected_nodes = std::move(nodes);
+    return result;
+  }
+
+  // The status of scheduling.
+  SchedulingResultStatus status;
+  // The nodes successfully scheduled.
+  std::vector<scheduling::NodeID> selected_nodes;
+};
+
 /// ISchedulingPolicy picks a node to from the cluster, according to the resource
 /// requirment as well as the scheduling options.
 class ISchedulingPolicy {
  public:
   virtual ~ISchedulingPolicy() = default;
   /// \param resource_request: The resource request we're attempting to schedule.
-  /// \param scheduling_options: scheduling options.
+  /// \param options: scheduling options.
   ///
   /// \return NodeID::Nil() if the task is unfeasible, otherwise the node id
   /// to schedule on.
   virtual scheduling::NodeID Schedule(const ResourceRequest &resource_request,
-                                      SchedulingOptions options) = 0;
+                                      SchedulingOptions options) {
+    return scheduling::NodeID();
```

Comment:
huh, should we by-default calling to batch scheduling policy? and vise versa?

Something like this:

```
class ISchedulingPolicy {
 virtual Schedule() = 0;
 virtual BatchSchedule() = 0;
}

class SingleSchedulePolicy : public ISchedulePolicy {
  Schedule() = 0;
  BatchSchedule() {
    for (req : ...) {
        Schedule(req)...
     }
  }
} 
class BatchSchedulePolicy : public ISchedulePolicy {
  Schedule() override {
      return BatchSchedule(...).result...;  
  }

  virtual BatchSchedule() = 0;
} 
```

Reply:
Nice ideal!
If so, we need to put the `seleted_nodes` into the context, because the single scheduling will rely on the `seleted_nodes` when doing batch scheduling. Like this:

```
struct SchedulingContext {
  virtual ~SchedulingContext() = default;
  // The nodes successfully scheduled.
  std::vector<scheduling::NodeID> selected_nodes_;
};

class ISingleSchedulePolicy : public ISchedulingPolicy {
 public:
  virtual ~ISingleSchedulePolicy() = default;

  SchedulingResult BatchSchedule(
      const std::vector<const ResourceRequest *> &resource_request_list,
      const SchedulingOptions &options,
      SchedulingContext *context = nullptr) override {
    for (auto &resource_request : resource_request_list) {
      auto result = Schedule(*resource_request, options, context);
      if (!result.status.IsSuccess()) {
        return result;
      }
      RAY_CHECK(result.selected_nodes.size() == 0);
      context->selected_nodes_.emplace_back(result.selected_nodes[0]);
    }

    return SchedulingResult::Success(std::move(context->selected_nodes_));
  }
};

class IBatchSchedulePolicy : public ISchedulingPolicy {
 public:
  virtual ~IBatchSchedulePolicy() = default;

  SchedulingResult Schedule(const ResourceRequest &resource_request,
                            const SchedulingOptions &options,
                            SchedulingContext *context = nullptr) override {
    return BatchSchedule({&resource_request}, options, context);
  }
};
```

Reply:
sgtm